keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,348 @@
+
+#### Should we prevent a new member with an inconsistent feature gate from joining the cluster?
+
+One way to prevent inconsistencies of feature gate between different members is to prevent a new member with an inconsistent feature gate from joining the cluster in the first place. 
+We do not think this is the right way because: 
+* we want to keep the ability to turn on/off a feature in a rolling sequence without bringing down the whole cluster. 
+
+
+### Risks and Mitigations
+
+#### Backport Risks
+
+To support feature gate as soon as in 3.6, we are backporting some proto changes into 3.5. There is the potential risk of changing a stable release. But we do not think this is a real risk, because it does not involve any changes other than adding a new proto field in an existing proto, and 3.5 server does not write or use this new field. Protos are inherently backward compatible. This change should also not affect upgrade/downgrade of 3.5.
+
+#### Data Compatibility Risks during Feature Value Change
+
+There are some features that would need to touch data in the apply/storage layer. Disabling them might result in orphaned data or unrecognizable data that could be breaking.
+How do we handle the data change? 
+1. if the etcd server knows this feature, it should be able to handle the data when the feature is on or off. How it handles the data at WAL entry `i` is determined by the cluster feature flag set from raft before WAL entry `i`. 
+1. if the etcd server does not know this feature, i.e. its version is not compatible with this feature, it means the data is not compatible with the storage version. The storage version migration should handle cleaning up older or newly introduced data.
+
+Overall, when developers introduce new data with a feature, they should be careful to include data cleaning in `schema.UnsafeMigrate`. But we prefer not to have a mechanism to disallow such features to be turned off after they are turned on because new features could be buggy, we need to keep the option to be able to turn it off without bring down the whole cluster.
+
+#### Feature Implementation Change Risks
+
+There are cases when a developer wants to change how an existing feature works. This would make things complicated in a mixed version cluster. One member could have the old implementation while another member uses the latest implementation. How can we make sure the cluster works consistently?
+
+Similar to [how Kubernetes handles feature gating changes for compatibility version](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/4330-compatibility-versions#feature-gating-changes), if the feature change affects data in anyway, we need to keep the implementation history for at least 2 minor versions (allowed version skew), like
+```go
```

Comment:
Just to clarify, we want to keep 2 implementations (if there's delta), not 2 previous minor versions (3 implementations), correct?

Also I wonder if we can make it easier for developers to make sure they keep the implementation. I am thinking some interface such as (**very roughly**):

```
type FeatureGateRunner struct {
  featuregate.Feature
  Handlers map[Version]func() error
}

func (r FeatureGateRunner) HandleAtVersion(v Version) error {
  return r.Handlers[v]()
}
```

Reply:
I don't think the additional interface is going to be beneficial. The main benefits of feature gate is it is easy to use, insert the following whereever I want to add new codes. 
```
if featureGate.Enabled(feature) {
  // my fancy new codes
}
```
With the wrapper, it is much cumbersome to use and remove in the future. 

I would expect change of feature direction to happen rarely. It is fine for it to be a little bit harder to keep multiple implementations.